packages/create-component-with-subscriptions/src/createComponentWithSubscriptions.js

```diff
@@ -0,0 +1,222 @@
+
+  // Unique state key name to avoid conflicts.
+  const stateWrapperKey = `____${property}`;
+
+  // If possible, extend the specified component to add subscriptions.
+  // This preserves ref compatibility and avoids the overhead of an extra fiber.
+  let BaseClass = (Component: any);
+  const prototype = (Component: any).prototype;
+
+  // If this is a functional component, use a HOC.
+  // Since functional components can't have refs, that isn't a problem.
+  // Class component lifecycles are required, so a class component is needed anyway.
+  if (
+    typeof prototype !== 'object' ||
+    typeof prototype.render !== 'function' ||
+    Component.____isSubscriptionHOC === true
+  ) {
+    BaseClass = class extends React.Component {
+      static ____isSubscriptionHOC = true;
+      render() {
+        const props = {
+          ...this.props,
+          [property]: this.state[property],
+        };
+        return React.createElement(Component, props);
+      }
+    };
+  }
+
+  // Event listeners are only safe to add during the commit phase,
+  // So they won't leak if render is interrupted or errors.
+  const subscribeHelper = (props, instance) => {
+    if (props[property] != null) {
+      const wrapper = instance.state[stateWrapperKey];
+
+      const valueChangedCallback = value => {
+        instance.setState(state => {
+          // If the value is the same, skip the unnecessary state update.
+          if (state[property] === value) {
+            return null;
+          }
+
+          const currentSubscribable =
+            state[stateWrapperKey] !== undefined
+              ? state[stateWrapperKey].subscribable
+              : null;
+
+          // If this event belongs to an old or uncommitted data source, ignore it.
+          if (wrapper.subscribable !== currentSubscribable) {
+            return null;
+          }
+
+          return {
+            [property]: value,
+          };
+        });
+      };
+
+      // Store subscription for later (in case it's needed to unsubscribe).
+      // This is safe to do via mutation since:
+      // 1) It does not impact render.
+      // 2) This method will only be called during the "commit" phase.
+      wrapper.subscription = subscribe(props, valueChangedCallback);
+
+      // External values could change between render and mount,
+      // In some cases it may be important to handle this case.
+      const value = getValue(props);
+      if (value !== instance.state[property]) {
+        instance.setState({
+          [property]: value,
+        });
+      }
+    }
+  };
+
+  const unsubscribeHelper = (props, instance) => {
+    if (props[property] != null) {
+      const wrapper = instance.state[stateWrapperKey];
+
+      unsubscribe(props, wrapper.subscription);
+
+      wrapper.subscription = null;
+    }
+  };
+
+  // Extend specified component class to hook into subscriptions.
+  class Subscribable extends BaseClass {
```

Comment:
I'm concerned about the inheritance pattern. I assume that this is for refs? Another issue with HoC might be devtools/tree size but minor.

Ref forwarding has come up a number of times. We should probably just bite the bullet and fix that.